src/accelerate/utils/bitsandbytes.py

```diff
@@ -0,0 +1,332 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import logging
+
+import torch
+import torch.nn as nn
+
+from .imports import (
+    is_4bit_bnb_available,
+    is_8bit_bnb_available,
+    is_bnb_available,
+    is_peft_available,
+    is_transformers_available,
+)
+
+
+if is_bnb_available():
+    import bitsandbytes as bnb
+
+logger = logging.getLogger(__name__)
+
+
+def get_bnb_model(model, bnb_config, is_peft_model):
+    """
+        This function will quantize the input model with the parameters specified in bnb_config.
+
+    Args:
+        model (`torch.nn.Module`): Input model
+        bnb_config (`dict`'): The bitsandbytes parameters
+
+    Returns:
+        `torch.nn.Module`: The quantized model
+    """
+    if is_transformers_available():
+        from transformers import BitsAndBytesConfig
+        from transformers.utils.bitsandbytes import get_keys_to_not_convert
+
+    quantization_config = BitsAndBytesConfig.from_dict(bnb_config, return_unused_kwargs=False)
+
+    load_in_4bit = quantization_config.load_in_4bit
+    load_in_8bit = quantization_config.load_in_8bit
+
+    if (load_in_4bit and not is_4bit_bnb_available()) or (load_in_8bit and not is_8bit_bnb_available()):
+        raise ImportError(
+            "You have a version of `bitsandbytes` that is not compatible with 4bit inference and training."
+            "Make sure you have the latest version of `bitsandbytes` installed"
+        )
+    # get dtype
+    # set dtype of the model disregarding what we have in config
+    torch_dtype = model.dtype
+    if torch_dtype != torch.float16:
+        # We force the `dtype` to be float16, this is a requirement from `bitsandbytes`
+        logger.info(
+            f"Overriding torch_dtype={torch_dtype} with `torch_dtype=torch.float16` due to "
+            "requirements of `bitsandbytes` to enable model loading in 8-bit or 4-bit. "
+            "Pass your own torch_dtype to specify the dtype of the remaining non-linear layers or pass"
+            " torch_dtype=torch.float16 to remove this warning."
+        )
+        torch_dtype = torch.float16
+        model.to(torch_dtype)
+
+    # some modules needs to be in fp32
+    keep_in_fp32_modules = model._keep_in_fp32_modules
+    if keep_in_fp32_modules is not None:
+        for name, param in model.named_parameters():
+            if any(module_to_keep_in_fp32 in name for module_to_keep_in_fp32 in keep_in_fp32_modules):
+                param = param.to(torch.float32)
```

Comment:
This is specific to Transformers and should be an argument of the function here.